### ISSUE TITLE
EdgeHub not stable

### ISSUE THREAD

**Issue by SarithaMaran** [Azure/iotedge#356]
# EdgeHub not stable

EdgeHub is not working properly if we keep the modules running for more than a day. 

## Modules description

I have a edge device with 4 modules

-  (1) A custom C# code (Generates some data similar to tempsensor)
-  (2) A custom C# code (Filters)
-  (3) Azure function C# code (Calculates something)
-  (4) ASA module (compares values of current and previous data)

The first module simply generates simulated data and publishes in every 1 min duration. The other modules process this data send to upstream.
And here is the route :
```{
  "routes": {
    "module2ToIoTHub": "FROM /messages/modules/module2/outputs/* INTO $upstream",
    "module1Tomodule2": "FROM /messages/modules/module1/outputs/output1 INTO BrokeredEndpoint(\"/modules/module2/inputs/input1\")",
    "module3ToIoTHub": "FROM /messages/modules/module3/outputs/* INTO $upstream",
    "module1Tomodule3": "FROM /messages/modules/module1/outputs/output1 INTO BrokeredEndpoint(\"/modules/module3/inputs/input1\")",
    "alertsToCloud": "FROM /messages/modules/module4/* INTO $upstream",
    "module1ToAsa": "FROM /messages/modules/module1/outputs/output1 INTO BrokeredEndpoint(\"/modules/module4/inputs/input1\")"
  }
}
```

## The Problem

- In the beginning everything works fine but when this modules run for more than 1 day the edgeHub stops working properly and data is not transmitted to other modules
- After some time, like 10 hours edgeHub shows the following logs:
```
2018-09-21 08:11:07.179 +00:00 [ERR] - Module RASPI_58/module3 is not connected
2018-09-21 08:11:07.179 +00:00 [ERR] - Module RASPI_58/module2 is not connected
2018-09-21 08:11:07.180 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-09-21 08:11:08.260 +00:00 [ERR] - Module RASPI_58/module2 is not connected
2018-09-21 08:11:08.260 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-09-21 08:11:08.260 +00:00 [ERR] - Module RASPI_58/module3 is not connected
2018-09-21 08:11:10.102 +00:00 [ERR] - Module RASPI_58/module2 is not connected
2018-09-21 08:11:10.290 +00:00 [ERR] - Module RASPI_58/module3 is not connected
2018-09-21 08:11:10.290 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-09-21 08:11:14.051 +00:00 [ERR] - Module RASPI_58/module3 is not connected
2018-09-21 08:11:14.052 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-09-21 08:11:14.251 +00:00 [ERR] - Module RASPI_58/module2 is not connected
2018-09-21 08:11:21.433 +00:00 [ERR] - Module RASPI_58/module2 is not connected
2018-09-21 08:11:22.292 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-09-21 08:11:22.721 +00:00 [ERR] - Module RASPI_58/module3 is not connected
2018-09-21 08:11:36.653 +00:00 [ERR] - Module RASPI_58/module2 is not connected
2018-09-21 08:11:39.904 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-09-21 08:11:41.031 +00:00 [ERR] - Module RASPI_58/module3 is not connected
```
- To recover I have to restart the edgeHub module and than also the other modules.
- But still the in between data will be discarded as edgehub is restarted.

## Messages are delayed in inter-module communication 

- As mentioned in the routes the first module i.e module1 forwards the data to module2 , module3 and also to module4. I have noticed that even if edgeHub is working properly after some hours the module3 is getting data delayed almost by one hour compared to module2.

## Expected Behavior
 EdgeHub should work proper with multiple modules and should route the data to modules at proper time 

## Current Behavior
 getting data delayed 

### Device (Host) Operating System
 Raspberry Pi Stretch  and ubuntu16.04

### Architecture
 arm32 and amd64 

### Container Operating System
 Linux containers 

#### iotedged
 Run `iotedge version` 

#### Edge Agent
 iotedge 1.0.1 

#### Edge Hub
 iotedge 1.0.1 

#### Docker
Client:
 Version:      18.06.0-dev
 API version:  1.37
 Go version:   go1.10.2
 Git commit:   daf021fe
 Built:        Wed Jun 27 00:53:52 2018
 OS/Arch:      linux/arm
 Experimental: false
 Orchestrator: swarm

## Additional Information
- And also I have noticed that sometimes the inter-module communication is proper. The modules work fine but no data could be sent to upstream i.e(IotHub)
- Facing the problem in both amd-64 and arm-32 architectures

**Comment by myagley** [comment]
Hello. Thank you for the issue report.

The `Module RASPI_58/module3 is not connected` error message is logged when a module is not connected to the Edge Hub to allow routing to the module.

Can you please help understand more about the various modules. What version of the IoT Hub SDK is being used? Based on the logs it looks like the modules are not reconnecting to the Edge Hub.

**Comment by SarithaMaran** [comment]
@myagley 
I have mentoined the version.
It shows in the logs version as 1.0.1.15962126.
And about the modules:
    module1- A custom C# code (Generates some data similar to tempsensor)
    module2- A custom C# code (Filters)
    module3- Azure function C# code (Calculates something)
    module4- ASA module (compares values of current and previous data)

And sometimes the edgeHub throws this error:
```
2018-10-01 06:57:21.450 +00:00 [ERR] - Module RASPI_58/module4 is not connected
2018-10-01 06:57:29.081 +00:00 [WRN] - Did not receive ack for message b68a5dca-1a73-4f32-a6c8-44d3909116f4 from device/module RASPI_58/module3
2018-10-01 06:57:29.100 +00:00 [WRN] - Error sending messages to module RASPI_58/module3
System.TimeoutException: Message completion response not received
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.SendMessageAsync(IMessage message, String input) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 236
   at Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint.ModuleMessageProcessor.<>c__DisplayClass5_2.<<ProcessAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/ModuleEndpoint.cs:line 98
2018-10-01 06:57:29.554 +00:00 [INF] - Opened link Events for RASPI_58/VesselStatusModule
2018-10-01 06:57:29.932 +00:00 [WRN] - Error disposing message in ModuleMessages link for RASPI_58/VesselStatusModule
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Devices.Client.InternalClient.<>c.<ApplyTimeout>b__61_1(Task t)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__278_1(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.<>c__DisplayClass26_0.<<InvokeFunc>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 132
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 112
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.ProcessMessageFeedbackAsync(String messageId, FeedbackStatus feedbackStatus) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 115
   at Microsoft.Azure.Devices.Edge.Hub.Amqp.LinkHandlers.SendingLinkHandler.DisposeMessageAsync(Delivery delivery) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/linkhandlers/SendingLinkHandler.cs:line 104

```

**Comment by myagley** [comment]
Is it possible to get the nuget version number of the C# SDK in your modules? It should be `1.17.0` or `1.18.0`. Reconnecting is handled by the modules themselves, so it is helpful to understand what is running there.

This might be only possible with your custom modules (module 1 and 2) as I don't think you will have insight into the ASA module.

**Comment by myagley** [comment]
The NRE fix was checked in here: #541

**Comment by myagley** [comment]
This has been fixed in the 1.0.5 release. More information on upgrading is here: https://docs.microsoft.com/en-us/azure/iot-edge/how-to-update-iot-edge
